Source/WTF/wtf/cocoa/UUIDCocoa.mm

```diff
@@ -35,11 +35,17 @@
     return [[NSUUID alloc] initWithUUIDString:toString()];
 }
 
-// Use string instead of bytes to avoid consideration of endianess (NSUUID stores UUID as array of bytes,
-// and we store UUID as UInt128).
-UUID::UUID(NSUUID * nsUUID)
-    : UUID(*parse(String([nsUUID UUIDString])))
+UUID::UUID(NSUUID *nsUUID)
+    : m_data(deletedValue)
```

Comment:
> Default initialization to the deleted value is probably incorrect. Usually we default initialize to the empty value. One thing that will be incorrect is operator bool, which checks for the empty value. And logically, a default UUID is not a deleted bucket in a hash table.

Sure, I will change it to empty value.

> Why is it OK for some operations to fail if the client happens to use the 0x0 or 0x1 UUID? (A client generating random UUIDs will never hit this case; but a client making them by hand might.) Many WebCore data types use special empty and deleted values, and we make that work by controlling all the values ourselves. But UUIDs can come from uncontrolled places, like public APIs.

0 is nil UUID (https://www.ietf.org/rfc/rfc4122.txt), so it's a special value by spec and might be okay to fail the operations. 
1 is should be a valid value; I think it might be unexpected to fail operations on it.

> What guarantees that all clients of the UUID class will know to check for validity before using

I think identifier classes (e.g. `ObjectIdentifier`) generally need to be validated before using?

> I think it might be better for the UUID class not to offer a hash table emptyValue or deletedValue. Instead, clients that want to put a UUID in a hash table should use a wrapper that can represent empty and deleted. (Maybe we already support this using optional or pair?)

Maybe we could introduce a new class RandomUUID for existing use cases, which has emptyValue and deletedValue (i.e. cannot 0 or 1) and can be used for hash table. And make UUID accept 0 and 1 (mapping to value in NSUUID) which cannot be used for hash table directly?
(I am not sure what you mean by supporting using optional or pair...)

This patch just fixes the bug that NSUUID of 0x1 is converted to a WebCore::UUID with random value (i.e. NSUUID -> WebCore::UUID -> different NSUUID). I think "whether 0x0 or 0x1 should be treated as valid value" is a different issue that needs more discussion.

Reply:
I think we should validate and reject deletedValue NSUUID. Instead of having a constructor, UUID::fromNSUUID should exist and it should return std::optional based on the value. And when 0x1 NSUUID comes, we should just reject it.

The deletedValue and emptyValue are not selected randomly in UUID value space. Since canonical UUID needs version header (see the spec of UUID), these two values are never valid UUID. UUID generation must not generate these values (otherwise, it is not a UUID). So from the input, when accepting it as WTF::UUID, we should reject them. If they are generated in a right manner, they are never 0x1. And NSUUID already rejects 0x0 UUID (it is also the invalid value). So it is natural that we reject the other invalid UUIDs.

I prefer current deletedValue / emptyValue largely because of memory reason. Having another boolean for HashMap / HashSet is too costly since UUID is 16-byte aligned: This means these boolean gets 16-byte size.

Reply:
BTW, we sometimes would like to use all the value range of the keys in HashMap. For that purpose, we already have GenericHashKey<T> wrapper. See some use cases in JSC. But for UUID, I think rejecting invalid UUID when adopting WTF::UUID looks a right path to me.

Reply:
Thanks! FWIW, GenericHashKey is what I was trying to find when I suggested

> I think it might be better for the UUID class not to offer a hash table emptyValue or deletedValue. Instead, clients that want to put a UUID in a hash table should use a wrapper that can represent empty and deleted. (Maybe we already support this using optional or pair?)

Given that @Constellation and @szewai have pointed out that both 0x0 and 0x1 are defined to be invalid UUIDs by spec, I like the suggestion to
* change UUID(NSUUID*) to UUID::fromNSUUID, and have it return optional (so we can check for invalid input)
* always fail UUID parsing if 0x0 or 0x1 is the input (prevents hash table corruption, clearer)
* keep using emptyValue 0x0 and deletedValue 0x1 (saves memory, simpler)

If there is a use case for 0x0 and 0x1 that we know we use, I would also be OK with a GenericHashKey solution or with having two classes, UUID and UUIDWithRFCIncompatibleBytes. But that seems more complicated.

Reply:
According to spec (https://www.rfc-editor.org/rfc/rfc4122#section-4.1.2) the last byte is Node, not version, so I am not sure why 0x1 is picked.
I agree a normal UUID generator will not generate 0x0 or 0x1 though (based on the rules specified in the spec), so we might just reject it. The proposal sounds good!

Reply:
> According to spec ([rfc-editor.org/rfc/rfc4122#section-4.1.2](https://www.rfc-editor.org/rfc/rfc4122#section-4.1.2)) the last byte is Node, not version, so I am not sure why 0x1 is picked.

I mean, UUID version number is 1-5. So if it is 0x1, it means version number is 0, thus, not a valid UUID.
https://www.rfc-editor.org/rfc/rfc4122#section-4.1.3

Reply:
I see. I think that means every value with version 0 should be treated as invalid (e.g. 0x2, 0x3, ...), but existing implementation actually allows those values, so the criteria might be confusing to users (who crafts NSUUID with special values): "why 1 is not valid in WebKit but 2 is"

Reply:
If it's clearer to be even more restrictive, I think that's an OK thing to do. Better to be restrictive up front. Once we start accepting UUIDs and putting them on disk, we probably can't go back.